### PR TITLE
Parquet bloom filter

### PR DESCRIPTION
**Pull request by cjjnjust**

### CODE REVIEW

File: parquet-column/src/main/java/org/apache/parquet/column/impl/ColumnWriterV1.java [hunk 1]
```diff
@@ -180,6 +199,7 @@ public void write(double value, int repetitionLevel, int definitionLevel) {
     definitionLevelColumn.writeInteger(definitionLevel);
     dataColumn.writeDouble(value);
     updateStatistics(value);
+    if (bloomData != null) bloomData.insert(value);
```
Comment (winningsix) [inline parquet-column/src/main/java/org/apache/parquet/column/impl/ColumnWriterV1.java]: code style: add {} around.

Comment (winningsix) [reply parquet-column/src/main/java/org/apache/parquet/column/impl/ColumnWriterV1.java]: You can use the property to check whether bloom filter is enabled.

File: parquet-column/src/main/java/org/apache/parquet/column/impl/ColumnWriterV1.java [hunk 2]
```diff
@@ -260,17 +289,21 @@ public void close() {
...
-    return repetitionLevelColumn.getAllocatedSize()
-    + definitionLevelColumn.getAllocatedSize()
-    + dataColumn.getAllocatedSize()
-    + pageWriter.allocatedSize();
+    long size = repetitionLevelColumn.getAllocatedSize()
+      + definitionLevelColumn.getAllocatedSize()
+      + dataColumn.getAllocatedSize()
+      + pageWriter.allocatedSize();
+
+    return bloomData == null ? size : size + bloomData.getAllocatedSize();
```
Comment (winningsix) [inline parquet-column/src/main/java/org/apache/parquet/column/impl/ColumnWriterV1.java]: can we have a variable to get the bloom filter data size?

File: parquet-column/src/main/java/org/apache/parquet/column/impl/ColumnWriterV2.java [hunk 3]
```diff
@@ -230,37 +258,45 @@ public void finalizeColumnChunk() {
       }
       dataColumn.resetDictionary();
     }
+
+    if (bloomDataWriter != null && bloomData != null) {
+      bloomData.flush();
```
Comment (winningsix) [inline parquet-column/src/main/java/org/apache/parquet/column/impl/ColumnWriterV2.java]: Why should we need to flush the bloom data here? We should flush the data after using bloom data.

File: parquet-column/src/main/java/org/apache/parquet/column/impl/ColumnWriterV2.java [hunk 4]
```diff
@@ -230,37 +258,45 @@ public void finalizeColumnChunk() {
...
    * @return the number of bytes of memory used to buffer the current data
    */
   public long getCurrentPageBufferedSize() {
-    return repetitionLevelColumn.getBufferedSize()
-        + definitionLevelColumn.getBufferedSize()
-        + dataColumn.getBufferedSize();
+    long size = repetitionLevelColumn.getBufferedSize()
+      + definitionLevelColumn.getBufferedSize()
+      + dataColumn.getBufferedSize();
+    return bloomData == null ? size : bloomData.getBufferedSize();
```
Comment (winningsix) [inline parquet-column/src/main/java/org/apache/parquet/column/impl/ColumnWriterV2.java]: The same as above.

File: parquet-column/src/main/java/org/apache/parquet/column/values/bloom/BloomDataWriter.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.parquet.column.values.bloom;
+
+public interface BloomDataWriter {
+  void writeBloomData(Bloom bloom);
```
Comment (winningsix) [inline parquet-column/src/main/java/org/apache/parquet/column/values/bloom/BloomDataWriter.java]: Java Doc needed.

File: parquet-column/src/main/java/org/apache/parquet/column/impl/ColumnWriteStoreV1.java [hunk 6]
```diff
@@ -61,7 +56,13 @@ public ColumnWriter getColumnWriter(ColumnDescriptor path) {
 
   private ColumnWriterV1 newMemColumn(ColumnDescriptor path) {
     PageWriter pageWriter = pageWriteStore.getPageWriter(path);
-    return new ColumnWriterV1(path, pageWriter, props);
+
+    if (props.isBloomFilterEnabled() && props.getBloomFilterColumnNames() != null) {
+      BloomDataWriter bloomDataWriter = pageWriteStore.getBloomDataWriter(path);
+      return new ColumnWriterV1(path, pageWriter, bloomDataWriter, props);
```
[the remark this answers is not in the document]
Comment (cjjnjust) [reply parquet-column/src/main/java/org/apache/parquet/column/impl/ColumnWriteStoreV1.java]: We need to get bloomDataWriter from pageWriteStore, not from props.

File: parquet-column/src/main/java/org/apache/parquet/column/impl/ColumnWriterV1.java (hunk 1 above)
Comment (cjjnjust) [reply parquet-column/src/main/java/org/apache/parquet/column/impl/ColumnWriterV1.java]: sometimes, user may enable the bloom filter, but not specify the bloom filter enabled column.
In this case, we should not create bloom filter.

File: parquet-column/src/main/java/org/apache/parquet/column/impl/ColumnWriterV2.java (hunk 3 above)
Comment (cjjnjust) [reply parquet-column/src/main/java/org/apache/parquet/column/impl/ColumnWriterV2.java]: the flush here means we set bit set.  when columnwriter finish write all values, we need to also set bits in bitset.

**Comment by cjjnjust** [comment]
Hi @winningsix 
Please see https://github.com/intel-hadoop/parquet-mr/pull/2

**Pull request closed without merging**